src/devices/cpu/z80/z80.cpp

```diff
@@ -3668,6 +3677,7 @@ void z80_device::execute_set_input(int inputnum, int state)
 		break;
 
 	case INPUT_LINE_IRQ0:
+		if(state && !m_irq_state) m_irq_at = total_cycles() + m_icount;
```

Comment:
execute_set_input doesn't guarantee that the irq will happen, Z80 IRQ flag might be disabled.
If for some reason spectrum.cpp needs to know the moment the interrupt occurs, it's doable by (ab)using irqack_cb() and then checking the cpu cycles in the driver.

Reply:
irqack_cb() is not the option per my understanding. We need to know how much HALT ate from current frame as it can't be interrupted.

Reply:
> This looks pretty much a driver responsibility, and we actually have an HALT callback in z80 core too. As it stands this PR needs reverting.

I wasn't able to make it work on driver side with couple attempts as we need to know how much uninteraptable HALT/NOP overlaps with next frame.
As now everything works well, I'll give another try.

Reply:
see: #9645